python/lsst/daf/butler/core/quantum.py

```diff
@@ -165,7 +131,7 @@ def initInputs(self) -> NamedKeyDict[DatasetType, DatasetRef]:
         with `DatasetType` instances as keys (names can also be used for
         lookups) and `DatasetRef` instances as values.
         """
-        return self._initInputs
+        return copy.copy(self._initInputs)
```

Comment:
Check out `NamedKeyDict.freeze`; I think it's exactly what you want here and in similar accessors (or maybe the constructor).  That'd come with changing the return type `NamedKeyMapping[DatasetType, DatasetRef]`.